### PR TITLE
[CK_TILE] Add fmha fwd N-Warp S-Shuffle pipeline (fmha fwd splitkv pipeline variant)

### PR DESCRIPTION
**Pull request by poyenc**
1. Add new fmha fwd splitkv pipeliene: **N-Warp** **S-Shuffle**. This pipeline was specificly otimized for small seqlen_q (<= 16)
    -  **N-Warp**: use block warps=<1, 4, 1> both for gemm0 & gemm1. Parallelize computation on `kN0` (seqlen_k) & `kN1` (hdim_v) direction
    -  **S-Shuffle**: use LDS to shuffle gemm0 output S to meet gemm1 input register layout requirement
2. Use more # warps in splitkv combine kernel. Decide the `kM0` tile size automatically
3. Add small warp gemms: 4x64x16 & 64x4x16 (implemented by 4x4x4 mfma intrinsics)

thanks to @qianfengz for helping opitmize the pipelines

### CODE REVIEW

**Comment by poyenc** [comment]
need to undo pipeline changes in fwd_splitkv.py

File: include/ck_tile/ops/gemm/warp/warp_gemm.hpp [hunk 1]
```diff
@@ -56,6 +56,24 @@ using WarpGemmMfmaF16F16F32M32N32K16SwizzleBTransposedCDistribution =
...
 
+using WarpGemmMfmaF16F16F32M4N64K16 = WarpGemmImpl<WarpGemmAtrributeMfmaIterateK<
+    WarpGemmAttributeMfmaImplF16F16F32M4N64K4<WGAttrCtlEnum::Default_>,
+    4>>;
+
+using WarpGemmMfmaF16F16F32M4N64K16TransposedCDistribution =
+    WarpGemmImpl<WarpGemmAtrributeMfmaIterateKAndTransposedCDistribution<
+        WarpGemmAttributeMfmaImplF16F16F32M4N64K4<WGAttrCtlEnum::Default_>,
+        4>>;
+
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm.hpp]: `TransposedCDistribution` is not suitable for M4N64K16 situation, should not be defined here

File: include/ck_tile/ops/gemm/warp/warp_gemm.hpp [hunk 2]
```diff
@@ -56,6 +56,24 @@ using WarpGemmMfmaF16F16F32M32N32K16SwizzleBTransposedCDistribution =
...
+using WarpGemmMfmaF16F16F32M64N4K16 = WarpGemmImpl<WarpGemmAtrributeMfmaIterateK<
+    WarpGemmAttributeMfmaImplF16F16F32M64N4K4<WGAttrCtlEnum::Default_>,
+    4>>;
+
+using WarpGemmMfmaF16F16F32M64N4K16TransposedCDistribution =
+    WarpGemmImpl<WarpGemmAtrributeMfmaIterateKAndTransposedCDistribution<
+        WarpGemmAttributeMfmaImplF16F16F32M64N4K4<WGAttrCtlEnum::Default_>,
+        4>>;
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm.hpp]: TransposedCDistribution is not suitable for M64N4K16 situation, should not be defined here

File: include/ck_tile/ops/gemm/warp/warp_gemm.hpp [hunk 3]
```diff
@@ -104,6 +122,24 @@ using WarpGemmMfmaBf16Bf16F32M32N32K16SwizzleBTransposedCDistribution =
...
         2>>;
 
+using WarpGemmMfmaBf16Bf16F32M4N64K16 = WarpGemmImpl<WarpGemmAtrributeMfmaIterateK<
+    WarpGemmAttributeMfmaImplBf16Bf16F32M4N64K4<WGAttrCtlEnum::Default_>,
+    4>>;
+
+using WarpGemmMfmaBf16Bf16F32M4N64K16TransposedCDistribution =
+    WarpGemmImpl<WarpGemmAtrributeMfmaIterateKAndTransposedCDistribution<
+        WarpGemmAttributeMfmaImplBf16Bf16F32M4N64K4<WGAttrCtlEnum::Default_>,
+        4>>;
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm.hpp]: TransposedCDistribution is not suitable for M4N64K16 situation, should not be defined here

File: include/ck_tile/ops/gemm/warp/warp_gemm.hpp [hunk 4]
```diff
@@ -104,6 +122,24 @@ using WarpGemmMfmaBf16Bf16F32M32N32K16SwizzleBTransposedCDistribution =
...
+
+using WarpGemmMfmaBf16Bf16F32M64N4K16 = WarpGemmImpl<WarpGemmAtrributeMfmaIterateK<
+    WarpGemmAttributeMfmaImplBf16Bf16F32M64N4K4<WGAttrCtlEnum::Default_>,
+    4>>;
+
+using WarpGemmMfmaBf16Bf16F32M64N4K16TransposedCDistribution =
+    WarpGemmImpl<WarpGemmAtrributeMfmaIterateKAndTransposedCDistribution<
+        WarpGemmAttributeMfmaImplBf16Bf16F32M64N4K4<WGAttrCtlEnum::Default_>,
+        4>>;
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm.hpp]: TransposedCDistribution is not suitable for M64N4K16 situation, should not be defined here

File: include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp [hunk 5]
```diff
@@ -29,6 +29,10 @@ template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float
 template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 16, 16, 16, true> { using Type = WarpGemmMfmaF16F16F32M16N16K16TransposedCDistribution; };
 template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 16, 16, 32, false> { using Type = WarpGemmMfmaF16F16F32M16N16K32; };
 template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 16, 16, 32, true> { using Type = WarpGemmMfmaF16F16F32M16N16K32TransposedCDistribution; };
+template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 4, 64, 16, false> { using Type = WarpGemmMfmaF16F16F32M4N64K16; };
+template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 4, 64, 16, true> { using Type = WarpGemmMfmaF16F16F32M4N64K16TransposedCDistribution; };
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp]: Avoid using the TransposedCDistribution for M4N64K16

File: include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp [hunk 6]
```diff
@@ -29,6 +29,10 @@ template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float
...
+template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 64, 4, 16, false> { using Type = WarpGemmMfmaF16F16F32M64N4K16; };
+template<> struct WarpGemmMfmaDispatcher<ck_tile::half_t, ck_tile::half_t, float, 64, 4, 16, true> { using Type = WarpGemmMfmaF16F16F32M64N4K16TransposedCDistribution; };
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp]: Avoid using the TransposedCDistribution for M64N4K16

File: include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp [hunk 7]
```diff
@@ -42,6 +46,10 @@ template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float
 template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 16, 16, 16, true> { using Type = WarpGemmMfmaBf16Bf16F32M16N16K16TransposedCDistribution; };
 template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 16, 16, 32, false> { using Type = WarpGemmMfmaBf16Bf16F32M16N16K32; };
 template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 16, 16, 32, true> { using Type = WarpGemmMfmaBf16Bf16F32M16N16K32TransposedCDistribution; };
+template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 4, 64, 16, false> { using Type = WarpGemmMfmaBf16Bf16F32M4N64K16; };
+template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 4, 64, 16, true> { using Type = WarpGemmMfmaBf16Bf16F32M4N64K16TransposedCDistribution; };
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp]: Avoid using the TransposedCDistribution for M64N4K16

File: include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp [hunk 8]
```diff
@@ -42,6 +46,10 @@ template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float
...
+template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 64, 4, 16, false> { using Type = WarpGemmMfmaBf16Bf16F32M64N4K16; };
+template<> struct WarpGemmMfmaDispatcher<ck_tile::bf16_t, ck_tile::bf16_t, float, 64, 4, 16, true> { using Type = WarpGemmMfmaBf16Bf16F32M64N4K16TransposedCDistribution; };
```
Comment (qianfengz) [inline include/ck_tile/ops/gemm/warp/warp_gemm_dispatcher.hpp]: Avoid using the TransposedCDistribution for M64N4K16

**Review by qianfengz — DISMISSED** [review]

**Review by qianfengz — APPROVED** [review]

**Pull request merged**